OneSignalSDK/onesignal/src/main/java/com/onesignal/OSNotificationWillShowInForegroundManager.java

```diff
@@ -0,0 +1,208 @@
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "extNotifJobKey is null or manager does not have a job with key: " + extNotifJobKey + ", returning Result.failure for Worker");
+                return Result.failure();
+            }
+
+            // Make sure current job obtained is not null
+            ExtNotificationGenerationJob currentExtNotifJob = (ExtNotificationGenerationJob) manager.getJob(extNotifJobKey);
+            if (currentExtNotifJob == null) {
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "currentExtNotifJob is null for key: " + extNotifJobKey + ", returning Result.failure for Worker");
+                return Result.failure();
+            }
+
+            // Make sure extNotificationWillShowInForegroundHandler is not null
+            if (OneSignal.extNotificationWillShowInForegroundHandler == null) {
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "OneSignal.extNotificationWillShowInForegroundHandler is null, returning Result.failure for Worker");
+                return Result.failure();
+            }
+
+            // Fire the ext handler and report success
+            OneSignal.onesignalLog(OneSignal.LOG_LEVEL.INFO, "Calling OneSignal.extNotificationWillShowInForegroundHandler.notificationWillShowInForeground(currentExtNotifJob) with OS notification id: " + currentExtNotifJob.getApiNotificationId());
+            OneSignal.extNotificationWillShowInForegroundHandler.notificationWillShowInForeground(currentExtNotifJob);
+
+            return Result.success();
+        }
+    }
+
+    public static class AppNotificationWillShowInForegroundWorker extends Worker {
+
+        public AppNotificationWillShowInForegroundWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
+            super(context, workerParams);
+        }
+
+        @NonNull
+        @Override
+        public Result doWork() {
+            OneSignal.onesignalLog(OneSignal.LOG_LEVEL.INFO, "doWork() called for AppNotificationWillShowInForegroundWorker");
+
+            // appContext my have been used to begin "doWork"
+            if (OneSignal.appContext == null) {
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "OneSignal.appContext is null, returning Result.failure for Worker");
+                return Result.failure();
+            }
+
+            // Make sure the key we need to to obtain our notifJob exists in the data input of the Worker
+            if (!getInputData().hasKeyWithValueOfType(APP_NOTIF_JOB_WORKER_KEY, String.class)) {
+                OneSignal.onesignalLog(OneSignal.LOG_LEVEL.ERROR, "getInputData() does not have the key: " + EXT_NOTIF_JOB_WORKER_KEY + " with class type: String, returning Result.failure for Worker");
+                return Result.failure();
+            }
+
+            // The job key must not be null and a job should exist for the key
+            String appNotifJobKey = getInputData().getString(APP_NOTIF_JOB_WORKER_KEY);
```

Comment:
We can do this in a follow up PR but we should just put a whole payload here instead of a key look. This way if the app process restarts we still have all the data to do the job.

Reply:
Changing to a worker that spans the processing work all the way to the notification showing instead